<?xml version="1.0" encoding="utf-8" ?> 
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>BiasViz 2</title>
<link href="style.css" type="text/css" rel="stylesheet" />
<link rel="icon" type="image/vnd.microsoft.icon" href="favicon.ico" />
</head>
<body>
<div id="content">
<h1>BiasViz 2: Protein Amino Acid Bias Analysis Applet</h1>
<p>by <a href="/huska">Matthew R. Huska</a></p>
<h2>Summary</h2>
<p>This applet can be used to look at the composition of amino acids within a sliding window given a multiple sequence alignment. The window size and amino acid(s) of interest can be modified and the results are displayed in real time, allowing the user to tweak both of these parameters to their liking. The values can also be downloaded as a CSV file which allows further visualization in a separate program such as Microsoft Excel.</p>

<div id="launchbox"><a id="launchtext" href="applet.html">Launch BiasViz 2</a>
<p class="note">* requires Java 1.5 or greater (<a href="http://www.java.com/download">download here</a>). You will also need to accept a certificate so that the program can load your input files from your computer.</p>
<p class="note">If you still have problems starting BiasViz because of security warnings, there is a <a href="http://www.java.com/en/download/help/jcp_security.xml">workaround here</a>. BiasViz needs permission to read files from your computer so that it can load your input files.</p>
</div>

<p>For further details please see the paper:</p>
<pre>
BiasViz: visualization of amino acid biased regions in protein alignments.
Huska MR, Buschmann H, Andrade-Navarro MA.
Bioinformatics. 2007 Nov 15;23(22):3093-4. Epub 2007 Oct 6.
PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/17921493">17921493</a>
Fulltext: <a href="http://dx.doi.org/10.1093/bioinformatics/btm489">doi:10.1093/bioinformatics/btm489</a>
</pre>

<h2>Examples</h2>
<p>Examples of three input file to be displayed simultaneously by BiasViz2.</p>

<a href="examples/hdNterm_fasta.txt">hdNterm_fasta.txt</a>
<p>
Multiple sequence alignment of the N-terminal of human Huntingtin with selected orthologs in FASTA format. This file and format can be obtained with programs like ClustalX using individual sequences as input.
</p>

<a href="examples/hdNterm_jpred.txt">hdNterm_jpred.txt</a>
<p>"Simple HTML" output file for secondary structure prediction obtained from Jpred3 using the advanced option and the FASTA file above as input.</p>

<a href="examples/hdNterm_ard2.txt">hdNterm_ard2.txt</a>
<p>File with values attached to all amino acids in the multiple sequence alignment obtained using the ARD2 web tool for alpha-solenoid prediction and the FASTA file above as input. Separators between sequences are those of the original FASTA file. Amino acid lines indicate position, amino acid, score, and other two values that we don't need in this case. For example, like this: "140,D,0.89,0,1". BiasViz will read only the first three values (the second one just for verification) and ignore the rest.</p>

<p>Questions, comments or bugs? Email me here: <a href="mailto: matthuska@alumni.uwaterloo.ca">matthuska@alumni.uwaterloo.ca</a></p>
<p class="note">Want to use the old version? Go here: <a href="http://www.ogic.ca/projects/biasviz/">BiasViz 1</a></p>
<div id="footer"><p>BiasViz is an open source project. Download the source code from <a href="http://sourceforge.net/projects/biasviz/">our Sourceforge page</a>.</p>
<a href="http://sourceforge.net/"><img src="http://sflogo.sourceforge.net/sflogo.php?group_id=200141&amp;type=2" width="125" height="37" alt="SourceForge.net Logo" /></a>
</div>
</div>
<!-- Google Analytics snippet -->
<script src='http://www.google-analytics.com/ga.js' type='text/javascript' />
<script type="text/javascript">
try{
var pageTracker = _gat._getTracker("UA-6628261-1");
pageTracker._trackPageview();
} catch(err) {}</script>
<!-- End Google Analytics snippet -->
</body>
</html>

